### ISSUE TITLE
Stream layer's reader seems to be closed before it is used

### ISSUE THREAD

**Issue by tliron** [google/go-containerregistry#707]
```go
import (
	"io"
	"io/ioutil"

	namepkg "github.com/google/go-containerregistry/pkg/name"
	"github.com/google/go-containerregistry/pkg/v1/empty"
	"github.com/google/go-containerregistry/pkg/v1/mutate"
	"github.com/google/go-containerregistry/pkg/v1/remote"
	"github.com/google/go-containerregistry/pkg/v1/stream"
)

func PushToRegistry(readCloser io.ReadCloser, name string) error {
	tag, err := namepkg.NewTag(name)
	if err != nil {
		return err
	}

	// This works
	layer := stream.NewLayer(ioutil.NopCloser(readCloser))

	// But this breaks :(
	//layer = stream.NewLayer(readCloser)

	image, err := mutate.AppendLayers(empty.Image, layer)
	if err != nil {
		return err
	}

	return remote.Write(tag, image)
}
```
If I run the code as is it works. But if I uncomment the "but this breaks" line instead I get errors like this:
```
Patch "http://192.168.39.201:5000/v2/hi/blobs/uploads/2ddf0b1a-89a8-4c75-aeaa-c946f18cc4fa?_state=yS_jHDbW8iEUHPOEJLdlsX_halshFZpjU2zphNb3eaB7Ik5hbWUiOiJoaSIsIlVVSUQiOiIyZGRmMGIxYS04OWE4LTRjNzUtYWVhYS1jOTQ2ZjE4Y2M0ZmEiLCJPZmZzZXQiOjAsIlN0YXJ0ZWRBdCI6IjIwMjAtMDQtMDVUMjM6Mjg6NTYuMjU1ODE3OTk4WiJ9": net/http: HTTP/1.x transport connection broken: close /myfile.txt: file already closed
```
Note that I am not closing the ReadCloser anywhere else.

Am I misunderstanding something?

**Comment by cppforlife** [comment]
i tracked down what the problem is since i was also getting `HTTP/1.x transport connection broken: close xxx: file already closed`. explained here: https://github.com/k14s/kbld/commit/fed37ec1e811ab6b315024101d4df38a834403e2#diff-008f24a3dbd1a6a2ab530b571ba259f9R76-R80. it might be reasonable to stop Closing blob in streamBlob since net/http is closing it as well.

**Comment by jonjohnsonjr** [comment]
Thanks for the investigation -- I believe my linked PR is the suggested fix, but let me know if I'm missing something.

**Comment by tliron** [comment]
It seems the fix has introduced a new bug. In v0.1.4 the code above results in different errors, like this:
```
GET http://localhost:5000/v2/catalog/hello/manifests/latest: MANIFEST_UNKNOWN: manifest unknown; map[Tag:latest]
```
Again, wrapping with the `NopCloser` makes this error go away and everything works as expected.

**Comment by jonjohnsonjr** [comment]
> GET http://localhost:5000/v2/catalog/hello/manifests/latest

It's surprising to see this from the above code -- we shouldn't be doing any manifest GETs there. Are you sure that's the issue?

**Comment by tliron** [comment]
@jonjohnsonjr You are correct, the error above is from later in my code. The actual error is the same as before. E.g.:
```
Patch "https://localhost:5000/v2/catalog/hello/blobs/uploads/5e824f39-04aa-4aad-8244-261fafb3044c?_state=cWjK_b8IDzDFTU5b9n4kN_H_x2qznE8X7tkomuhnmW97Ik5hbWUiOiJjYXRhbG9nL2hlbGxvIiwiVVVJRCI6IjVlODI0ZjM5LTA0YWEtNGFhZC04MjQ0LTI2MWZhZmIzMDQ0YyIsIk9mZnNldCI6MCwiU3RhcnRlZEF0IjoiMjAyMC0xMC0yMlQyMjowNzowNy40MDU2NDQxMTVaIn0%3D": net/http: HTTP/1.x transport connection broken: close /spool/catalog\hello.txt: file already closed
```
In other words, it doesn't look like this fix worked, at least not for me.

**Comment by jonjohnsonjr** [comment]
I can't reproduce it, but I'll leave this open for now.

**Comment by tliron** [comment]
@jonjohnsonjr Could you share the code you wrote to reproduce it? I want to help get to the bottom of this.

**Comment by jonjohnsonjr** [comment]
@tliron you were right -- see the attached PR for a fix.

**Comment by tliron** [comment]
Thank you! It seems the PR can't be merged yet, I will be happy to test it once it is ready.

By the way, thank you for your work on this project. I have just spun my own work building on this into a new project, hope it will be of interest and use: [Reposure](https://github.com/tliron/reposure).

**Comment by tliron** [comment]
Confirmed resolved in release 0.3.0. Thank you!
